monix-eval/shared/src/main/scala/monix/eval/instances/CatsEffectForTask.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2014-2017 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.eval.instances
+
+import cats.effect.{Effect, IO}
+import monix.eval.{Callback, Task}
+import monix.execution.Scheduler
+
+/** Cats type class instances for [[monix.eval.Task Task]]
+  * for  `cats.effect.Effect` (and implicitly for
+  * `Applicative`, `Monad`, `MonadError`, `Sync`, etc).
+  *
+  * Note this is a separate class from [[CatsAsyncForTask]], because we 
+  * need an implicit [[monix.execution.Scheduler Scheduler]] in scope 
+  * in order to trigger the execution of a `Task`. However we cannot
+  * inherit directly from `CatsAsyncForTask`, because it would create 
+  * conflicts due to that one having a higher priority but being a 
+  * super-type.
+  *
+  * References:
+  *
+  *  - [[https://typelevel.org/cats/ typelevel/cats]]
+  *  - [[https://github.com/typelevel/cats-effect typelevel/cats-effect]]
+  */
+class CatsEffectForTask(implicit ec: Scheduler)
+  extends CatsBaseForTask with Effect[Task] {
+
+  override def runAsync[A](fa: Task[A])
+    (cb: Either[Throwable, A] => IO[Unit]): IO[Unit] = {
+
+    import CatsEffectForTask.noop
+    IO(fa.runAsync(new Callback[A] {
+      def onSuccess(value: A): Unit =
+        cb(Right(value)).unsafeRunAsync(noop)
+      def onError(ex: Throwable): Unit =
+        cb(Left(ex)).unsafeRunAsync(noop)
+    }))
+  }
+
+  /** We need to mixin [[CatsAsyncForTask]], because if we
```

Comment:
Is this true even with the priority of the traits mixed into Task? Does this also mean you could have conflicts for Monad (etc) which both CatsEffectForTask and CatsAsyncForTask implement?

Reply:
Yes, because unfortunately Scala's implicit resolution is a complex thing that nobody really knows how it works.

But the gist is this — the compiler basically calculates a score. And so this works:

```scala
class Foo
class Bar extends Foo

trait Level1 extends Level0 {
  implicit bar: Bar = new Bar
}

trait Level0 {
  implicit foo: Foo = new Foo
}
```

But this does not:

```scala
trait Level1 extends Level0 {
  implicit foo: Foo = new Foo
}

trait Level0 {
  implicit bar: Bar = new Bar
}
```

This is because `Bar <: Foo` and somehow the compiler when calculating the score of those 2 implicits will end up with the same number, the different cake levels being balanced against the subtyping relationship we have here.

Fortunately what I did works, even for `Monad`:

```scala
scala> import monix.execution.Scheduler.Implicits.global
import monix.execution.Scheduler.Implicits.global

scala> import monix.eval._
import monix.eval._

scala> implicitly[cats.Monad[Task]]
res0: cats.Monad[monix.eval.Task] = monix.eval.instances.CatsAsyncForTask$@2f600d06
```

This is because the compiler does not see `CatsEffectForTask` as being a subtype of `CatsAsyncForTask`. A weird rule, but there you have it.